api/v1/harborcluster_types.go

```diff
@@ -16,25 +16,321 @@ limitations under the License.
 package v1
 
 import (
+	cmmeta "github.com/jetstack/cert-manager/pkg/apis/meta/v1"
+	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
 // EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
 // NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
 
+// the name of component used harbor cluster.
+type Component string
+
+// all Component used in harbor cluster full stack.
+const (
+	ComponentHarbor   Component = "harbor"
+	ComponentRedis    Component = "redis"
+	ComponentStorage  Component = "storage"
+	ComponentDatabase Component = "database"
+)
+
 // HarborClusterSpec defines the desired state of HarborCluster
 type HarborClusterSpec struct {
 	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
 	// Important: Run "make" to regenerate code after modifying this file
 
-	// Foo is an example field of HarborCluster. Edit HarborCluster_types.go to remove/update
-	Foo string `json:"foo,omitempty"`
+	// harbor version to be deployed, this version determines the image tags of harbor service components
+	// +kubebuilder:validation:Required
+	// https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string
+	// +kubebuilder:validation:Pattern="^(?P<major>0|[1-9]\\d*)\\.(?P<minor>0|[1-9]\\d*)\\.(?P<patch>0|[1-9]\\d*)(?:-(?P<prerelease>(?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\\.(?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\\+(?P<buildmetadata>[0-9a-zA-Z-]+(?:\\.[0-9a-zA-Z-]+)*))?$"
+	Version string `json:"version"`
+
+	// The url exposed to clients to access harbor
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Pattern="^https?://.*$"
+	PublicURL string `json:"publicURL"`
+
+	// Password for the root admin
+	// +kubebuilder:validation:Required
+	AdminPasswordSecret string `json:"adminPasswordSecret"`
+
+	// Secret reference for the TLS certs
+	// +optional
+	TLSSecre string `json:"tlsSecret"`
+
+	// The issuer for Harbor certificates.
+	// If the 'kind' field is not set, or set to 'Issuer', an Issuer resource
+	// with the given name in the same namespace as the Certificate will be used.
+	// If the 'kind' field is set to 'ClusterIssuer', a ClusterIssuer with the
+	// provided name will be used.
+	// The 'name' field in this stanza is required at all times.
+	CertificateIssuerRef cmmeta.ObjectReference `json:"certificateIssuerRef"`
+
+	// Indicates that the harbor is paused.
+	// +optional
+	Paused bool `json:"paused,omitempty"`
```

Comment:
Should this paused be in spec? What case for

Reply:
This one is inherited from the core operator and it's also a standard K8s spec item.